### ISSUE TITLE
Understanding "moving" folder in ClickHouse data directory

### ISSUE THREAD

**Issue by Vasu7052** [ClickHouse/ClickHouse#49098]
Hi Team, 

We wanted to know the significance of the "moving" folder in ClickHouse data directory:  `/data-dir/clickhouse/store/595/5955e84b-7082-45fe-9955-e84b7082d5fe/moving/*`

Reason we wanted to know about this is due to high disk space consumption by this folder. (as 70% of the disk is utilized only by this folder).

Also, Can we delete the content of this folder? If yes, then what is the best way to do so?

**Comment by CheSema** [comment]
I suppose you are using settings `move_factor`. [Doc](https://clickhouse.com/docs/en/engines/table-engines/mergetree-family/mergetree#table_engine-mergetree-multiple-volumes_configure).
If you do, then the moves happen regularly in your setup.

It seems like here is a bug.

`cloned_part = parts_mover.clonePart(moving_part);`
https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/MergeTree/MergeTreeData.cpp#L7708
There is a place when CH copies data from the first disk to the second in `moving` directory. After part is copied CH moves the part to working dir and enables it as active part in table.

While CH is copying a part, the part might become inactive, it might been merged into a bigger one, for example.
The code is ready for that.
https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/MergeTree/MergeTreePartsMover.cpp#L263
Please, check how many such rows in logs do you have?
`Failed to swap {}. Active part doesn't exist. Possible it was merged or mutated. Will remove copy on path '{}'.`
I suspect that CH leave the part on the disk in that case instead removing it as useless.

Actually such scenarios when CH leaves some part with data on disk is unavoidable. For example, due to power outage.
In order to collect such useless pats, CH has special background process.
`clearOldTemporaryDirectories`
https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/MergeTree/MergeTreeData.cpp#L2006
And I do not see that it clears parts from moving directory.

So, in short:
- CH has to drop part when move has been canceled
- CH has to gather the useless parts from `[MOVING_DIR_NAME](https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/MergeTree/MergeTreeData.h#L156)`  directory.

**Comment by Vasu7052** [comment]
@CheSema thank you for a detailed answer. Few more think i want to clear up.

1. We are not using the `move_factor` setting as of now. Here is our storage.xml file for your reference:
```<yandex>
 <storage_configuration>
   <disks>
     <fast>
        <keep_free_space_bytes>1024</keep_free_space_bytes>
	<path>/ssd-disk/clickhouse/</path>
     </fast>
     <slow>
         <keep_free_space_bytes>1024</keep_free_space_bytes>
	 <path>/hdd-disk/clickhouse/</path>
     </slow>
   </disks>
   <policies>
      <hot_cold_data>
       <volumes>
         <hot_data_volume>
            <disk>fast</disk>
         </hot_data_volume>
         <cold_data_volume>
            <disk>slow</disk>
         </cold_data_volume>
       </volumes>
      </hot_cold_data>
   </policies>
 </storage_configuration>
</yandex>```

2. As a workaround we are thinking of running this bash command periodically to delete old folders in "moving" folder that were modified before 24 hours: 
`find /ssd-disk/clickhouse/store/12c/12cd9ee8-c629-4c25-aead-9ee8c6293c25/moving/ -maxdepth 1 -mindepth 1 -type d -mtime +0 -exec rm -rf {} ;`
Our hypothesis is that if the folder is not modified in last 24 hours, then the Clickhouse might got failed removing this particular folder. so until permanent fix is done in Clickhouse repo, can we run this command after also checking with `lsof`?

3. When we are running normal deletes or lightweight deletes from HDD disk only, size of SSD disk increases exponentially and we noticed this SSD space was increasing due to "moving" folder only. so is it normal or we are missing some setting in our config?

**Comment by CheSema** [comment]
1.
move_factor -- (by default, 0.1) -- from the [doc](https://clickhouse.com/docs/en/engines/table-engines/mergetree-family/mergetree#table_engine-mergetree-multiple-volumes_configure). 
So as soon as you have 2 volumes, parts are moving from first to second. Unless you set move_factor = 1.0.

Also move could be triggered by TTL MOVE. [Doc](https://clickhouse.com/docs/en/engines/table-engines/mergetree-family/mergetree#mergetree-table-ttl)
`TO DISK|VOLUME`

2.
For me 24 is Ok, but the data is yours. Just make calculation how fast it can be copied from disk to disk for sure. However `lsof` gives good enough safety check, use it with combination of time.

3.
Well. Thanks for highlighting that parts are leaved on src disk. Honestly it is surprising. In my hypothesis, second disk might have useless parts.
Do you have some loads with either
[MOVE PARTITION|PART](https://clickhouse.com/docs/en/sql-reference/statements/alter/partition#move-partitionpart)
or
[MOVE PARTITION TO TABLE](https://clickhouse.com/docs/en/sql-reference/statements/alter/partition#move-partition-to-table)?
Or may be some TTL rules with moving?

**Comment by Vasu7052** [comment]
@CheSema Apologies for missing out the TTL information. The following is the TTL condition we have applied to our table, which is based on a DateTime column called insertion_time:
`TTL insertion_time TO VOLUME 'hot_data_volume', insertion_time + toIntervalWeek(4) TO VOLUME 'cold_data_volume'`

So is there a clean way to delete data from `cold_data_volume` i.e. data older than 4 weeks (with some extra conditions) without getting size of `hot_data_volume`(our SSD disk) increased by its "moving" folder?

**Comment by CheSema** [comment]
@Vasu7052
Thanks for TTL definition
`TTL insertion_time TO VOLUME 'hot_data_volume', insertion_time + toIntervalWeek(4) TO VOLUME 'cold_data_volume'`

It explain why hot_data_volume is used as destination in move.
`TTL insertion_time TO VOLUME 'hot_data_volume'`
You explicitly say here that all data need to be moved to  hot_data_volume.

Let me explain. The new data is written on first suitable volume, it is  hot_data_volume in your case. And after that 2 option:

**Comment by den-crane** [comment]
>TTL insertion_time TO VOLUME 'hot_data_volume' -- it is redundant, delete that part.

Actually it's harmful because it moves data from cold to hot disk. Which should never happen, because parts should be moved only in direction automatically.

It can create state when parts move back and forth because `move_factor` moves parts to cold and immediately TTL moves the same part back.

**Comment by CheSema** [comment]
> It can create a state when parts move back and forth because move_factor moves parts to cold and immediately TTL moves the same parts back.

I do not follow, why? Sounds scary. 
For such case, ttl predicate has to become `true` after part was moved to cold. I do not see tre reason why ttl predicate could change the state after move operation.

**Comment by playdaddy** [comment]
In case someone ends up here and for some reasons uses old version (like we do)
In our case we had same situation with moving directory but because of another reason.
Had relatively large part (~500G) and CH wasn't fast enough during move, so ongoing mutations constantly was changing part version. After move CH was checking versions, see difference and do nothing
Had to stop mutations to give it some time

**Comment by den-crane** [comment]
> > It can create a state when parts move back and forth because move_factor moves parts to cold and immediately TTL moves the same parts back.
> 
> I do not follow, why? Sounds scary. For such case, ttl predicate has to become `true` after part was moved to cold. I do not see tre reason why ttl predicate could change the state after move operation.

It's really easy to reproduce.
You probably didn't understand that there is two TTL rules:
 
```
1. insertion_time TO VOLUME 'hot_data_volume', 
2. insertion_time + toIntervalWeek(4) TO VOLUME 'cold_data_volume'
```

The first rule is harmful, because it will move parts which satisfy the first rule from the cold disk to the hot disk.
I observed this issue with clients many times. Like they running out of a disk space on the cold disk, to fix it they use `alter ... move to cold_data_volume` for a couple oldest parts, to free some disk space on the hot disk, and then Clickhouse moves these parts back to satisfy the first rule.
If you remove the first rule, nothing bad happens. Because inserts by default go to the first disk (which is `hot`).

**Comment by jetgm** [comment]
@CheSema When I upgraded ClickHouse from 22.8 to 23.8, the database startup process was very slow. Upon checking the logs, I found that two large tables were consistently performing remove operations on moving directory files, and the directory contained over ten thousand subdirectories. My system is configured with a tiered storage strategy similar to the one described in this post, without any TTL rules. How can I avoid this situation from occurring?

**Comment by den-crane** [comment]
@jetgm it's one time operation https://github.com/ClickHouse/ClickHouse/issues/49197

**Comment by jetgm** [comment]
> @jetgm it's one time operation #49197

got it. Thanks
